### DISCUSSION TITLE
CUDA error 209 no kernel image is available for execution on the device

### DISCUSSION THREAD

**Discussion by yonglie** [milvus-io/milvus#15746] [Q&A and General discussion]
Faiss assertion 'err__ == cudaSuccess' failed in void faiss::gpu::runL2Norm(faiss::gpu::Tensor<T, 2, true, IndexType>&, bool, faiss::gpu::Tensor<float, 1, true, IndexType>&, bool, cudaStream_t) [with T = float; TVec = float4; IndexType = int; cudaStream_t = CUstream_st*] at gpu/impl/L2Norm.cu:292; details: CUDA error 209 no kernel image is available for execution on the device

这是怎么回事呢？我已经有显卡了，还是报错
what's wrong? can you tell me

**Comment by yhmo** [comment]
Please read this page: https://en.wikipedia.org/wiki/CUDA

Milvus 1.x is built on CUDA 11.0:

- CUDA SDK 1.0 support for compute capability 1.0 – 1.1 (Tesla)[[29]](https://en.wikipedia.org/wiki/CUDA#cite_note-29)
- CUDA SDK 1.1 support for compute capability 1.0 – 1.1+x (Tesla)
- CUDA SDK 2.0 support for compute capability 1.0 – 1.1+x (Tesla)
- CUDA SDK 2.1 – 2.3.1 support for compute capability 1.0 – 1.3 (Tesla)[[30]](https://en.wikipedia.org/wiki/CUDA#cite_note-30)[[31]](https://en.wikipedia.org/wiki/CUDA#cite_note-31)[[32]](https://en.wikipedia.org/wiki/CUDA#cite_note-32)[[33]](https://en.wikipedia.org/wiki/CUDA#cite_note-33)
- CUDA SDK 3.0 – 3.1 support for compute capability 1.0 – 2.0 (Tesla, Fermi)[[34]](https://en.wikipedia.org/wiki/CUDA#cite_note-34)[[35]](https://en.wikipedia.org/wiki/CUDA#cite_note-35)
- CUDA SDK 3.2 support for compute capability 1.0 – 2.1 (Tesla, Fermi)[[36]](https://en.wikipedia.org/wiki/CUDA#cite_note-36)
- CUDA SDK 4.0 – 4.2 support for compute capability 1.0 – 2.1+x (Tesla, Fermi, more?).
- CUDA SDK 5.0 – 5.5 support for compute capability 1.0 – 3.5 (Tesla, Fermi, Kepler).
- CUDA SDK 6.0 support for compute capability 1.0 – 3.5 (Tesla, Fermi, Kepler).
- CUDA SDK 6.5 support for compute capability 1.1 – 5.x (Tesla, Fermi, Kepler, Maxwell). Last version with support for compute capability 1.x (Tesla).
- CUDA SDK 7.0 – 7.5 support for compute capability 2.0 – 5.x (Fermi, Kepler, Maxwell).
- CUDA SDK 8.0 support for compute capability 2.0 – 6.x (Fermi, Kepler, Maxwell, Pascal). Last version with support for compute capability 2.x (Fermi) (Pascal GTX 1070Ti Not Supported).
- CUDA SDK 9.0 – 9.2 support for compute capability 3.0 – 7.2 (Kepler, Maxwell, Pascal, Volta) (Pascal GTX 1070Ti Not Supported. CUDA SDK 9.0 and support CUDA SDK 9.2).
- CUDA SDK 10.0 – 10.2 support for compute capability 3.0 – 7.5 (Kepler, Maxwell, Pascal, Volta, Turing). Last version with support for compute capability 3.x (Kepler). 10.2 is the last official release for macOS, as support will not be available for macOS in newer releases.
- CUDA SDK 11.0 support for compute capability 3.5 – 8.0 (Kepler (in part), Maxwell, Pascal, Volta, Turing, Ampere (in part)).[[37]](https://en.wikipedia.org/wiki/CUDA#cite_note-37) New data types: [Bfloat16](https://en.wikipedia.org/wiki/Bfloat16) and TF32 on third-generations Tensor Cores.[[38]](https://en.wikipedia.org/wiki/CUDA#cite_note-38)
- CUDA SDK 11.1 – 11.6 support for compute capability 3.5 – 8.6 (Kepler (in part), Maxwell, Pascal, Volta, Turing, Ampere).[[](https://en.wikipedia.org/wiki/CUDA#cite_note-39)

According to this page, the CUDA 11.0 can support compute capability 3.5 – 8.0.

The NVIDIA TITAN series GPU cards from RTX 3050 to RTX 3090 is in compute capability 8.6. You will get CUDA error 209 if your GPU card is in this range.

**Reply by yonglie** [reply]
我可以买这种显卡吗？
GeForce GTX 2080Ti
我是在官网文档看到的

另外，请问您
我用索引IVFSQ8查询100多万条的2048维度的向量，感觉好慢，要20秒，不知道为什么，那个服务器是cpu的milvus

**Reply by yonglie** [reply]
另外我还想问下您，显卡的算力跟显卡的驱动有关吗？

 CUDA Driver Version / Runtime Version          11.6 / 10.1
  CUDA Capability Major/Minor version number:    8.6

我的驱动是11.6算力8.6

是不是换低版本的驱动就可以了呢？

**Reply by yhmo** [reply]
> 我可以买这种显卡吗？ GeForce GTX 2080Ti 我是在官网文档看到的
> 
> 另外，请问您 我用索引IVFSQ8查询100多万条的2048维度的向量，感觉好慢，要20秒，不知道为什么，那个服务器是cpu的milvus

GeForce GTX 2080Ti 的compute capability是7.5，应该是可以的。


IVFSQ8查询100多万条的2048维，正常耗时应该在10毫秒左右。性能不正常的原因有几个：
1. 第一次查询要从存储系统读取数据到内存，所以第一次查询会很耗时
2. server_config.yaml里的cache_size那项设置缓存大小不够，导致频繁从存储系统交换数据。cache_size默认大小是4GB，2048维100多万数据量大约是8GB ～ 16GB，如果索引未完成建好，默认缓存大小是不够放的。

**Reply by yhmo** [reply]
> 另外我还想问下您，显卡的算力跟显卡的驱动有关吗？
> 
> CUDA Driver Version / Runtime Version 11.6 / 10.1 CUDA Capability Major/Minor version number: 8.6
> 
> 我的驱动是11.6算力8.6
> 
> 是不是换低版本的驱动就可以了呢？

这个问题我也说不好，驱动对显卡的性能是有一定影响，但显卡核心是无法改变的，也许跟低版本的驱动也不兼容。
